As a user, I want my evaluations using the WRDS RFC forecast service to run without encountering an NPE

The declaration is below. I'll share the data in the first comment. The exception is this:

```
2025-03-06T18:28:28.405+0000  [pool-247-thread-1] INFO EvaluationService - The evaluation failed with the following stack trace: 
wres.pipeline.InternalWresException: Could not complete project execution
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:407)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:179)
	at wres.Functions.evaluate(Functions.java:135)
	at wres.server.EvaluationService.lambda$startEvaluation$2(EvaluationService.java:695)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: wres.pipeline.WresProcessingException: Encountered an error while processing evaluation 'Irvwdpwj_d8mWzpkDX6YNfJun2o': 
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:842)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:378)
	... 7 common frames omitted
Caused by: wres.io.ingesting.IngestException: An ingest task could not be completed.
	at wres.io.ingesting.SourceLoader.load(SourceLoader.java:165)
	at wres.pipeline.Evaluator.evaluate(Evaluator.java:600)
	... 8 common frames omitted
Caused by: java.util.concurrent.CompletionException: wres.reading.ReadException: While attempting to acquire a time-series from WRDS AHPS.
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1773)
	... 3 common frames omitted
Caused by: wres.reading.ReadException: While attempting to acquire a time-series from WRDS AHPS.
	at wres.reading.ReaderUtilities.getTimeSeries(ReaderUtilities.java:834)
	at wres.reading.wrds.ahps.WrdsAhpsReader.lambda$getTimeSeriesSupplier$1(WrdsAhpsReader.java:331)
	at java.base/java.util.stream.StreamSpliterators$InfiniteSupplyingSpliterator$OfRef.tryAdvance(StreamSpliterators.java:1358)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$0(StreamSpliterators.java:292)
	at java.base/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(StreamSpliterators.java:206)
	at java.base/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(StreamSpliterators.java:161)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:298)
	at java.base/java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
	at wres.io.ingesting.database.DatabaseTimeSeriesIngester.ingest(DatabaseTimeSeriesIngester.java:223)
	at wres.io.ingesting.SourceLoader.lambda$readAndIngestData$4(SourceLoader.java:397)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	... 5 common frames omitted
Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at wres.reading.ReaderUtilities.getTimeSeries(ReaderUtilities.java:814)
	... 15 common frames omitted
Caused by: java.lang.NullPointerException: null
	at java.base/java.util.Objects.requireNonNull(Objects.java:209)
	at wres.reading.wrds.ahps.WrdsAhpsJsonReader.readFromStream(WrdsAhpsJsonReader.java:131)
	at wres.reading.wrds.ahps.WrdsAhpsJsonReader.read(WrdsAhpsJsonReader.java:118)
	at wres.reading.wrds.ahps.WrdsAhpsReader.lambda$getTimeSeriesTuple$2(WrdsAhpsReader.java:373)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	... 3 common frames omitted
```

Thanks,

Hank

=====================

```yaml
label: Hank Repost RFC_AHPS_Flow_90day_CSV
observed:
  label: USGS NWIS Streamflow Observations
  sources:
    uri: https://nwis.waterservices.usgs.gov/nwis/iv
    interface: usgs nwis
  variable: "00060"
  type: observations
predicted:
  label: RFC_AHPS_Flow_90day_CSV
  sources:
    uri: http://WRDS/api/rfc_forecast/v2.0/forecast/streamflow
    interface: wrds ahps
  variable: QR
  type: single valued forecasts
features:
- observed: '01631000'
  predicted: FROV2
unit: ft3/s
reference_dates:
  minimum: 2024-02-02T00:00:00Z
  maximum: 2025-02-05T00:00:00Z
valid_dates:
  minimum: 2024-02-02T00:00:00Z
  maximum: 2025-02-05T00:00:00Z
lead_times:
  minimum: 0
  maximum: 240
  unit: hours
lead_time_pools:
  period: 6
  frequency: 6
  unit: hours
metrics:
  - root mean square error normalized
  - mean square error
  - bias fraction
  - sample size
  - root mean square error
  - mean square error skill score normalized
  - mean absolute error
  - coefficient of determination
  - mean error
  - volumetric efficiency
```

Triggered the desired pathway with a unit test (that currently fails):

```
org.opentest4j.AssertionFailedError: Unexpected exception thrown: wres.reading.ReadException: While attempting to acquire a time-series from WRDS AHPS.

	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:152)
	at org.junit.jupiter.api.AssertDoesNotThrow.createAssertionFailedError(AssertDoesNotThrow.java:84)
	at org.junit.jupiter.api.AssertDoesNotThrow.assertDoesNotThrow(AssertDoesNotThrow.java:75)
	at org.junit.jupiter.api.AssertDoesNotThrow.assertDoesNotThrow(AssertDoesNotThrow.java:58)
	at org.junit.jupiter.api.Assertions.assertDoesNotThrow(Assertions.java:3259)
	at wres.reading.wrds.ahps.WrdsAhpsReaderTest.testReadProducesHttp404ErrorAndNoNullPointerException(WrdsAhpsReaderTest.java:1186)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: wres.reading.ReadException: While attempting to acquire a time-series from WRDS AHPS.
	at wres.reading.ReaderUtilities.getTimeSeries(ReaderUtilities.java:834)
	at wres.reading.wrds.ahps.WrdsAhpsReader.lambda$getTimeSeriesSupplier$1(WrdsAhpsReader.java:331)
	at java.base/java.util.stream.StreamSpliterators$InfiniteSupplyingSpliterator$OfRef.tryAdvance(StreamSpliterators.java:1358)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
	at java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
	at wres.reading.wrds.ahps.WrdsAhpsReaderTest.lambda$testReadProducesHttp404ErrorAndNoNullPointerException$0(WrdsAhpsReaderTest.java:1187)
	at org.junit.jupiter.api.AssertDoesNotThrow.assertDoesNotThrow(AssertDoesNotThrow.java:71)
	... 6 more
Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at wres.reading.ReaderUtilities.getTimeSeries(ReaderUtilities.java:814)
	... 19 more
Caused by: java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:208)
	at wres.reading.wrds.ahps.WrdsAhpsJsonReader.readFromStream(WrdsAhpsJsonReader.java:131)
	at wres.reading.wrds.ahps.WrdsAhpsJsonReader.read(WrdsAhpsJsonReader.java:118)
	at wres.reading.wrds.ahps.WrdsAhpsReader.lambda$getTimeSeriesTuple$2(WrdsAhpsReader.java:382)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

The line:

```
	at wres.reading.wrds.ahps.WrdsAhpsJsonReader.readFromStream(WrdsAhpsJsonReader.java:131)
```

Next, fix the bug and make it pass.

--- a/wres-reading/src/wres/reading/wrds/ahps/WrdsAhpsReader.java
+++ b/wres-reading/src/wres/reading/wrds/ahps/WrdsAhpsReader.java
@@ -369,10 +369,20 @@ public class WrdsAhpsReader implements TimeSeriesReader
         return this.getExecutor()
                    .submit( () -> {
                        // Get the input stream and read from it
-                       try ( InputStream inputStream = WrdsAhpsReader.getByteStreamFromUri( dataSource.getUri() );
-                             Stream<TimeSeriesTuple> seriesStream = AHPS_READER.read( dataSource, inputStream ) )
+                       try ( InputStream inputStream = WrdsAhpsReader.getByteStreamFromUri( dataSource.getUri() ) )
                        {
-                           return seriesStream.toList(); // Terminal
+                           if ( Objects.isNull( inputStream ) )
+                           {
+                               LOGGER.warn( "Failed to obtain time-series data from {}. Returning an empty stream.",
+                                            dataSource.getUri() );
+
+                               return List.of();
+                           }
+
+                           try ( Stream<TimeSeriesTuple> seriesStream = AHPS_READER.read( dataSource, inputStream ) )
+                           {
+                               return seriesStream.toList(); // Terminal
+                           }
                        }
                    } );
     }

--- a/wres-reading/test/wres/reading/wrds/ahps/WrdsAhpsReaderTest.java
+++ b/wres-reading/test/wres/reading/wrds/ahps/WrdsAhpsReaderTest.java
@@ -9,10 +9,10 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
-import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.mockito.Mockito;
@@ -829,7 +829,7 @@ class WrdsAhpsReaderTest
         try ( Stream<TimeSeriesTuple> tupleStream = reader.read( fakeSource ) )
         {
             List<TimeSeries<Double>> actual = tupleStream.map( TimeSeriesTuple::getSingleValuedTimeSeries )
-                                                         .collect( Collectors.toList() );
+                                                         .toList();
 
             Geometry geometry = MessageUtilities.getGeometry( FEATURE_NAME,
                                                               "Front Royal",
@@ -1123,4 +1123,68 @@ class WrdsAhpsReaderTest
 
     }
 
+    @Test
+    void testReadProducesHttp404ErrorAndNoNullPointerException()
+    {
+        // GitHub issue #444
+
+        this.mockServer.when( HttpRequest.request()
+                                         .withPath( FORECAST_PATH_V2 )
+                                         .withMethod( GET ) )
+                       .respond( HttpResponse.notFoundResponse() );
+
+        URI fakeUri = URI.create( "http://localhost:"
+                                  + this.mockServer.getLocalPort()
+                                  + FORECAST_PATH_V2 );
+
+        Source fakeDeclarationSource = SourceBuilder.builder()
+                                                    .uri( fakeUri )
+                                                    .sourceInterface( SourceInterface.WRDS_AHPS )
+                                                    .build();
+
+        Dataset fakeDataset = DatasetBuilder.builder()
+                                            .sources( List.of( fakeDeclarationSource ) )
+                                            .build();
+
+        DataSource fakeSource = DataSource.of( DataDisposition.JSON_WRDS_AHPS,
+                                               fakeDeclarationSource,
+                                               fakeDataset,
+                                               Collections.emptyList(),
+                                               fakeUri,
+                                               DatasetOrientation.RIGHT,
+                                               null );
+
+        SystemSettings systemSettings = Mockito.mock( SystemSettings.class );
+        Mockito.when( systemSettings.getMaximumWebClientThreads() )
+               .thenReturn( 6 );
+        Mockito.when( systemSettings.getPoolObjectLifespan() )
+               .thenReturn( 30_000 );
+
+        TimeInterval interval = TimeIntervalBuilder.builder()
+                                                   .minimum( Instant.parse( "2018-01-01T00:00:00Z" ) )
+                                                   .maximum( Instant.parse( "2021-01-01T00:00:00Z" ) )
+                                                   .build();
+        Geometry geometry = Geometry.newBuilder()
+                                    .setName( FEATURE_NAME )
+                                    .build();
+        GeometryTuple geometryTuple = GeometryTuple.newBuilder()
+                                                   .setLeft( geometry )
+                                                   .build();
+        Features features = FeaturesBuilder.builder()
+                                           .geometries( Set.of( geometryTuple ) )
+                                           .build();
+        EvaluationDeclaration declaration =
+                EvaluationDeclarationBuilder.builder()
+                                            .validDates( interval )
+                                            .features( features )
+                                            .build();
+
+        WrdsAhpsReader reader = WrdsAhpsReader.of( declaration, systemSettings );
+
+        try ( Stream<TimeSeriesTuple> tupleStream = reader.read( fakeSource ) )
+        {
+            Assertions.assertDoesNotThrow( () -> tupleStream.map( TimeSeriesTuple::getSingleValuedTimeSeries )
+                                                            .toList() );
+        }
+    }
 }

